nexus/src/db/datastore.rs

```diff
@@ -1028,6 +1033,276 @@ impl DataStore {
+                .await,
+        )
+        .map_err(|e| public_error_from_diesel_pool(e, ErrorHandler::Server))?;
+        if range.is_some() {
+            return Err(Error::InvalidRequest {
+                message:
+                    "IP Pool cannot be deleted while it contains IP ranges"
+                        .to_string(),
+            });
+        }
+
+        // Delete the pool, conditional on the rcgen not having changed. This
+        // protects the delete from occuring if clients created a new IP range
+        // in between the above check for children and this query.
+        let now = Utc::now();
+        let updated_rows = diesel::update(dsl::ip_pool)
+            .filter(dsl::time_deleted.is_null())
+            .filter(dsl::id.eq(authz_pool.id()))
+            .filter(dsl::rcgen.eq(db_pool.rcgen))
+            .set(dsl::time_deleted.eq(now))
+            .execute_async(self.pool_authorized(opctx).await?)
+            .await
+            .map_err(|e| {
+                public_error_from_diesel_pool(
+                    e,
+                    ErrorHandler::NotFoundByResource(authz_pool),
+                )
+            })?;
+
+        if updated_rows == 0 {
+            return Err(Error::InvalidRequest {
+                message: "deletion failed due to concurrent modification"
+                    .to_string(),
+            });
+        }
+        Ok(())
+    }
+
+    pub async fn ip_pool_update(
+        &self,
+        opctx: &OpContext,
+        authz_pool: &authz::IpPool,
+        updates: IpPoolUpdate,
+    ) -> UpdateResult<IpPool> {
+        use db::schema::ip_pool::dsl;
+        opctx.authorize(authz::Action::Modify, authz_pool).await?;
+        diesel::update(dsl::ip_pool)
+            .filter(dsl::id.eq(authz_pool.id()))
+            .filter(dsl::time_deleted.is_null())
+            .set(updates)
+            .returning(IpPool::as_returning())
+            .get_result_async(self.pool_authorized(opctx).await?)
+            .await
+            .map_err(|e| {
+                public_error_from_diesel_pool(
+                    e,
+                    ErrorHandler::NotFoundByResource(authz_pool),
+                )
+            })
+    }
+
+    pub async fn ip_pool_list_ranges(
+        &self,
+        opctx: &OpContext,
+        authz_pool: &authz::IpPool,
+        pag_params: &DataPageParams<'_, std::net::IpAddr>,
+    ) -> ListResultVec<IpPoolRange> {
+        use db::schema::ip_pool_range::dsl;
+        opctx.authorize(authz::Action::ListChildren, authz_pool).await?;
+        let limit = pag_params.limit.get().into();
+        let res = if let Some(last_seen_addr) = pag_params.marker {
+            let addr = ipnetwork::IpNetwork::from(*last_seen_addr);
+            dsl::ip_pool_range
+                .filter(dsl::ip_pool_id.eq(authz_pool.id()))
+                .filter(dsl::time_deleted.is_null())
+                .filter(dsl::first_address.gt(addr))
```

Comment:
FWIW, I tried checking out this branch locally, and making this change.

- Diesel complained that `std::net::IpAddr` didn't implement `Expression`. This makes sense; all the other invocations of `paginated` use `DataPageParams` with a marker type that implements `ToSql`.
- I tried switching from `std::net::IpAddr` to `ipnetwork::IpNetwork`. Due to the implementation of `DataPageParams`, I had to do this all that way up in `http_entrypoints.rs`:


```rust
          let marker = match query.page {                                                                                                                                                                                                     
              WhichPage::First(_) => None,                                                                                                                                                                                                    
              WhichPage::Next(ref last_address) => Some(ipnetwork::IpNetwork::from(*last_address)),                                                                                                                                           
          };                                                                                                                                                                                                                                  
          let pag_params = DataPageParams {                                                                                                                                                                                                   
              limit: rqctx.page_limit(&query)?,                                                                                                                                                                                               
              direction: PaginationOrder::Ascending,                                                                                                                                                                                          
              marker: marker.as_ref(),                                                                                                                                                                                                        
          };                    
```

This let me add the following here:

```rust
      pub async fn ip_pool_list_ranges(                                                                                                                                                                                                       
          &self,                                                                                                                                                                                                                              
          opctx: &OpContext,                                                                                                                                                                                                                  
          authz_pool: &authz::IpPool,                                                                                                                                                                                                         
~~        pag_params: &DataPageParams<'_, ipnetwork::IpNetwork>,                                                                                                                                                                              
      ) -> ListResultVec<IpPoolRange> {                                                                                                                                                                                                       
          use db::schema::ip_pool_range::dsl;                                                                                                                                                                                                 
          opctx.authorize(authz::Action::ListChildren, authz_pool).await?;                                                                                                                                                                    
~~                                                                                                                                                                                                                                            
~~        paginated(dsl::ip_pool_range, dsl::first_address, pag_params)                                                                                                                                                                       
~~            .filter(dsl::ip_pool_id.eq(authz_pool.id()))                                                                                                                                                                                    
~~            .filter(dsl::time_deleted.is_null())                                                                                                                                                                                            
~~            .select(IpPoolRange::as_select())                                                                                                                                                                                               
~~            .load_async::<IpPoolRange>(self.pool_authorized(opctx).await?)                                                                                                                                                                  
~~            .await                                                                                                                                                                                                                          
~~            .map_err(|e| {                                                                                                                                                                                                                  
~~                public_error_from_diesel_pool(                                                                                                                                                                                              
~~                    e,                                                                                                                                                                                                                      
~~                    ErrorHandler::NotFoundByResource(authz_pool),                                                                                                                                                                           
~~                )                                                                                                                                                                                                                           
~_            })                                                                                                                                                                                                                              
      } 
```

Reply:
I'll give it a whirl. It's a long story, but I think I tried it originally but ran into some trait bound issues. I think those no longer apply, so this might work. And yes, ascending order is hardcoded. I could probably relax that, I'll check the DB indexes are still used. Thanks!